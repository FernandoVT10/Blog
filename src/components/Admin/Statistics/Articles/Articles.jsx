import ArticleCard from "./ArticleCard";
import Api from "../../../../ApiController";

import { useRouter } from "next/router";
import { useState, useEffect } from "react";

import "./Articles.scss";

export default ({ setShowArticlesTable }) => {
    const [monthViews, setMonthViews] = useState([]);
    const [dayViews, setDayViews] = useState([]);

    const router = useRouter();

    useEffect(() => {
        Promise.all([
            Api.get("views/getMostViewedArticles/month/5", true),
            Api.get("views/getMostViewedArticles/day/5", true)
        ]).then(([monthViews, dayViews]) => {
            setDayViews(dayViews);
            setMonthViews(monthViews);
        });
    }, []);

    const showArticlesTable = (sort) => {
        setShowArticlesTable(true);

        router.push({
            pathname: router.pathname,
            query: { sort }
        });
    };

    return (
        <div className="statistics-articles">
            <ArticleCard
            title="Most Viewed Article This Month"
            showArticlesTable={ () => showArticlesTable("month") }
            data={monthViews}/>
            <ArticleCard
            title="Most Viewed Article This Day"
            showArticlesTable={ () => showArticlesTable("day") }
            data={dayViews}/>
        </div>
    );
};