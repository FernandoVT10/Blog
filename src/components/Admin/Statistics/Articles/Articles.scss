@use "../../../../styles/variables";

.statistics-articles {
    display: flex;

    &__card {
        width: calc(50% - 40px);
        background: variables.$primary;

        &:nth-child(1) { margin-right: 40px; }

        &-content {
            padding: 15px;
            padding-bottom: 0;
        }

        &-label {
            display: block;
            color: white;
        }

        &-views {
            display: block;
            color: white;
            font-weight: bold;
            font-size: 20px;

            &--loading {
                display: block;
                width: 100px;
                height: 30px;
                background: variables.$black;
            }
        }

        &-cover {
            margin-top: 10px;
            object-fit: cover;
            width: 100%;
            height: 200px;

            &--loading {
                display: block;
                margin-top: 10px;
                width: 100%;
                height: 200px;
                background: variables.$black;
            }
        }

        &-title {
            color: white;
            font-weight: 600;
            font-size: 22px;
            margin: 20px 15px;

            &--loading {
                margin: 20px 15px;
                display: block;
                width: 80%;
                height: 35px;
                background: variables.$black;
            }
        }
    }

    &__articles-list {
        &-item {
            position: relative;
            display: flex;
            align-items: center;
            padding: 5px 0;

            &:nth-child(2n+1) {
                background: variables.$black;
            }
        }

        &-cover {
            width: 80px;
            height: 50px;
            object-fit: cover;
        }

        &-title {
            margin-left: 15px;
            color: white;
            font-weight: bold;
            font-size: 16px;
        }

        &-views {
            position: absolute;
            right: 15px;
            color: white;
            font-weight: bold;
            font-size: 18px;
        }

        @media screen and (max-width: 576px) {
            &-item {
                padding: 20px 0;
            }

            &-cover {
                display: none;
            }
        }
    }

    &__card-footer {
        display: flex;
        justify-content: center;
        align-items: center;
        padding: 15px;
        background: variables.$black;

        &-text {
            cursor: pointer;
            color: variables.$secondary;
            font-weight: bold;
        }
    }

    @media screen and (max-width: 992px) {
        flex-direction: column;

        &__card {
            width: 100%;
            margin: 0;
            
            &:nth-child(1) { margin-bottom: 40px; }
        }
    }
}

.holi-4 {
    display: inline-block;
    width: 80px;
    height: 50px;
    background: variables.$primary;
}

.holi-5 {
    display: inline-block;
    width: 50%;
    height: 30px;
    margin-left: 15px;
    background: variables.$primary;
}

.holi-6 {
    display: inline-block;
    position: absolute;
    right: 15px;
    width: 20%;
    height: 35px;
    margin-left: 15px;
    background: variables.$primary;
}

.statistics-articles__articles-list-item:nth-child(2n) {
    .holi-4,
    .holi-5,
    .holi-6 {
        background: variables.$black;
    }
}