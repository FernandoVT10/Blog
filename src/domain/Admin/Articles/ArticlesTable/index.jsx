import Articles from "./ArticlesTable";

export default Articles;