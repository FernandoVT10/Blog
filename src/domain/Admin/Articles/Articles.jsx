import ArticlesTable from "./ArticlesTable"
import EditArticle from "./ArticleEditor/EditArticle";
import AddArticle from "./ArticleEditor/AddArticle";
import Categories from "./Categories";

import { useEffect, useState } from "react";
import { useRouter } from "next/router";

export default () => {
    const [articleId, setArticleId] = useState("");
    const [addArticle, setAddArticle] = useState(false);

    const router = useRouter();

    useEffect(() => {
        const { editArticleId } = router.query;

        if(editArticleId) {
            setArticleId(editArticleId);
        } else {
            setArticleId();
        }
    }, [router.query]);

    if(articleId) {
        return <EditArticle articleId={articleId}/>;
    }

    if(addArticle) {
        return <AddArticle setAddArticle={setAddArticle}/>;
    }

    return (
        <div>
            <Categories/>
            <ArticlesTable setAddArticle={setAddArticle}/>
        </div>
    );
}