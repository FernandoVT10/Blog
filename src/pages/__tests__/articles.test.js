import { act } from "react-dom/test-utils";
import { render } from "react-dom";
import Articles from "../articles";

const CATEGORIES_MOCK = [
    { _id: 2, name: "Technology" },
    { _id: 6, name: "Landscapes" },
    { _id: 19, name: "Travels" }
];

const RESPONSE_MOCK = {
    articles: [
        {
            _id: 2,
            title: "Test article",
            description: "Test description",
            cover: "article-1.jpg"
        },
        {
            _id: 3,
            title: "Test article 2",
            description: "Test description 2",
            cover: "article-2.jpg"
        },
        {
            _id: 4,
            title: "Test article 3",
            description: "Test description 3",
            cover: "article-3.jpg"
        }
    ],
    pagination: {
        hasPrevPage: true,
        prevPage: 5,
        hasNextPage: true,
        nextPage: 7,
        pages: [
            { number: 2, active: false },
            { number: 3, active: false },
            { number: 4, active: false },
            { number: 5, active: false },
            { number: 6, active: true },
            { number: 7, active: false },
            { number: 8, active: false }
        ]
    }
};

let container;

describe('Articles page', () => {
    beforeEach(() => {
        fetchMock.doMock();

        fetchMock
            .once(JSON.stringify(CATEGORIES_MOCK))
            .once(JSON.stringify(RESPONSE_MOCK));

        container = document.createElement('div');
        document.body.appendChild(container);
    });
    
    afterEach(() => {
        document.body.removeChild(container);
        container = null;
    });

    it('Check if renders correctly', async () => {
        Object.defineProperty(window, 'location', {
            value: {
                search: "?categories=Landscapes&search=Test&offset=2"
            }
        });

        await act(async () => {
            render(<Articles/>, container);
        });

        expect(container).toMatchSnapshot();
    });
});