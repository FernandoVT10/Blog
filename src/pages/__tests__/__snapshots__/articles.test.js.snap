// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Articles page Check if renders correctly 1`] = `
<div>
  <div>
    <div
      class="full-screen-loader "
    >
      <div
        class="full-screen-loader__spinner spinner-border text-light"
        role="status"
      >
        <span
          class="sr-only"
        >
          Loading...
        </span>
      </div>
      <h3
        class="full-screen-loader__text"
      >
        Loading...
      </h3>
    </div>
    <header>
      <nav
        class="navbar navbar-expand-md navbar-dark"
      >
        <a
          class="navbar-brand navbar__brand"
          href="/"
        >
          <img
            alt="Icon"
            class="d-inline-block align-top rounded-circle"
            height="50"
            src="/favicon.ico"
            width="50"
          />
          <h1
            class="navbar__title"
          >
            Fernando Vaca Tamayo
          </h1>
        </a>
        <button
          aria-controls="main-navbar"
          aria-expanded="false"
          aria-label="Toggle navigation"
          class="navbar-toggler"
          data-target="#main-navbar"
          data-toggle="collapse"
          type="button"
        >
          <span
            class="navbar-toggler-icon"
          />
        </button>
        <div
          class="collapse navbar-collapse navbar__items"
          id="main-navbar"
        >
          <ul
            class="navbar-nav"
          >
            <li
              class="nav-item"
            >
              <a
                class="nav-link navbar__link"
                href="/"
              >
                Home
              </a>
            </li>
            <li
              class="nav-item"
            >
              <a
                class="nav-link navbar__link"
                href="/articles"
              >
                Articles
              </a>
            </li>
            <li
              class="nav-item"
            >
              <a
                class="nav-link navbar__link"
                href="/projects"
              >
                Projects
              </a>
            </li>
            <li
              class="nav-item"
            >
              <a
                class="nav-link navbar__link"
                href="/contact-me"
              >
                Contact me
              </a>
            </li>
          </ul>
        </div>
      </nav>
    </header>
    <div
      class="body"
    >
      <div
        class="articles-container container-fluid"
      >
        <div
          class="row mt-4"
        >
          <div
            class="col-12"
          >
            <div
              class="row articles-container__header"
            >
              <div
                class="col-10 col-sm-6 col-lg-8"
              >
                <h2
                  class="title articles-container__title"
                >
                  Articles
                </h2>
              </div>
              <div
                class="col-2 col-sm-6 col-lg-4"
              >
                <div
                  class="articles-filter"
                >
                  <form>
                    <div
                      class="articles-filter__search-container"
                    >
                      <input
                        autocomplete="search"
                        class="formulary__input"
                        placeholder="Search article"
                        type="search"
                        value=""
                      />
                      <button
                        class="articles-filter__search-button"
                        type="submit"
                      >
                        <i
                          aria-hidden="true"
                          class="fas fa-search"
                        />
                      </button>
                    </div>
                    <div
                      class="articles-filter__filter "
                    >
                      <input
                        autocomplete="search"
                        class="formulary__input articles-filter__filter-search"
                        placeholder="Search article"
                        type="search"
                        value=""
                      />
                      <div
                        class="articles-filter__categories-container"
                      >
                        <h4
                          class="articles-filter__categories-title"
                        >
                          Categories
                        </h4>
                        <div
                          class="aticles-filter__categories"
                        >
                          <div
                            class="articles-filter__category"
                          >
                            <input
                              class="articles-filter__checkbox"
                              id="category-Technology"
                              type="checkbox"
                            />
                            <label
                              class="articles-filter__label"
                              for="category-Technology"
                            >
                              <span
                                class="articles-filter__label-checkbox"
                              />
                              Technology
                            </label>
                          </div>
                          <div
                            class="articles-filter__category"
                          >
                            <input
                              checked=""
                              class="articles-filter__checkbox"
                              id="category-Landscapes"
                              type="checkbox"
                            />
                            <label
                              class="articles-filter__label"
                              for="category-Landscapes"
                            >
                              <span
                                class="articles-filter__label-checkbox"
                              />
                              Landscapes
                            </label>
                          </div>
                          <div
                            class="articles-filter__category"
                          >
                            <input
                              class="articles-filter__checkbox"
                              id="category-Travels"
                              type="checkbox"
                            />
                            <label
                              class="articles-filter__label"
                              for="category-Travels"
                            >
                              <span
                                class="articles-filter__label-checkbox"
                              />
                              Travels
                            </label>
                          </div>
                        </div>
                        <button
                          class="submit-button articles-filter__set-categories-button"
                          type="submit"
                        >
                          Set Categories
                        </button>
                      </div>
                    </div>
                    <a
                      class="articles-filter__filter-button "
                      href="#"
                    >
                      <i
                        aria-hidden="true"
                        class="fas fa-sort-amount-down-alt"
                      />
                    </a>
                  </form>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="col-12 col-lg-6 mt-4"
          >
            <div
              class="article-card"
            >
              <img
                alt="Test article"
                class="article-card__cover"
                src="/img/articles/article-1.jpg"
              />
              <div
                class="article-card__details"
              >
                <h3
                  class="article-card__title"
                >
                  Test article
                </h3>
                <p
                  class="article-card__description"
                >
                  Test description
                </p>
                <a
                  class="article-card__link"
                  href="/article/2"
                >
                  Continue Reading
                  <i
                    aria-hidden="true"
                    class="fas fa-arrow-right ml-1"
                  />
                </a>
              </div>
            </div>
          </div>
          <div
            class="col-12 col-lg-6 mt-4"
          >
            <div
              class="article-card"
            >
              <img
                alt="Test article 2"
                class="article-card__cover"
                src="/img/articles/article-2.jpg"
              />
              <div
                class="article-card__details"
              >
                <h3
                  class="article-card__title"
                >
                  Test article 2
                </h3>
                <p
                  class="article-card__description"
                >
                  Test description 2
                </p>
                <a
                  class="article-card__link"
                  href="/article/3"
                >
                  Continue Reading
                  <i
                    aria-hidden="true"
                    class="fas fa-arrow-right ml-1"
                  />
                </a>
              </div>
            </div>
          </div>
          <div
            class="col-12 col-lg-6 mt-4"
          >
            <div
              class="article-card"
            >
              <img
                alt="Test article 3"
                class="article-card__cover"
                src="/img/articles/article-3.jpg"
              />
              <div
                class="article-card__details"
              >
                <h3
                  class="article-card__title"
                >
                  Test article 3
                </h3>
                <p
                  class="article-card__description"
                >
                  Test description 3
                </p>
                <a
                  class="article-card__link"
                  href="/article/4"
                >
                  Continue Reading
                  <i
                    aria-hidden="true"
                    class="fas fa-arrow-right ml-1"
                  />
                </a>
              </div>
            </div>
          </div>
        </div>
        <div
          class="row mt-3"
        >
          <div
            class="col-12"
          >
            <div
              class="pagination"
            >
              <a
                class="pagination__arrow"
                href="#"
              >
                <i
                  aria-hidden="true"
                  class="fas fa-chevron-left"
                />
              </a>
              <ul
                class="pagination__number-container"
              >
                <li
                  class="pagination__number-item"
                >
                  <a
                    class="pagination__number "
                    href="#"
                  >
                    2
                  </a>
                </li>
                <li
                  class="pagination__number-item"
                >
                  <a
                    class="pagination__number "
                    href="#"
                  >
                    3
                  </a>
                </li>
                <li
                  class="pagination__number-item"
                >
                  <a
                    class="pagination__number "
                    href="#"
                  >
                    4
                  </a>
                </li>
                <li
                  class="pagination__number-item"
                >
                  <a
                    class="pagination__number "
                    href="#"
                  >
                    5
                  </a>
                </li>
                <li
                  class="pagination__number-item"
                >
                  <a
                    class="pagination__number pagination__number--active"
                    href="#"
                  >
                    6
                  </a>
                </li>
                <li
                  class="pagination__number-item"
                >
                  <a
                    class="pagination__number "
                    href="#"
                  >
                    7
                  </a>
                </li>
                <li
                  class="pagination__number-item"
                >
                  <a
                    class="pagination__number "
                    href="#"
                  >
                    8
                  </a>
                </li>
              </ul>
              <a
                class="pagination__arrow"
                href="#"
              >
                <i
                  aria-hidden="true"
                  class="fas fa-chevron-right"
                />
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
    <footer
      class="main-footer container-fluid mt-4"
    >
      <div
        class="full-screen-loader "
      >
        <div
          class="full-screen-loader__spinner spinner-border text-light"
          role="status"
        >
          <span
            class="sr-only"
          >
            Loading...
          </span>
        </div>
        <h3
          class="full-screen-loader__text"
        >
          Loading...
        </h3>
      </div>
      <div
        class="modal fade"
        id="-modal"
        role="dialog"
        tabindex="-1"
      >
        <div
          class="modal-dialog "
          role="document"
        >
          <div
            class="modal-content modal__content"
          >
            <div
              class="modal-header modal__header"
            >
              <h5
                class="modal__title"
              >
                Suscription confirm status
              </h5>
              <button
                aria-label="Close"
                class="close modal__button"
                data-dismiss="modal"
                type="button"
              >
                <span
                  aria-hidden="true"
                >
                  ×
                </span>
              </button>
            </div>
            <div
              class="modal-body modal__body"
            >
              <p />
            </div>
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-12 col-sm-6 col-xl-3"
        >
          <a
            class="main-footer__brand"
            href="/"
          >
            <img
              alt="Icon"
              class="d-inline-block align-top rounded-circle"
              height="50"
              src="/favicon.ico"
              width="50"
            />
            <h3
              class="main-footer__title"
            >
              Fernando Vaca Tamayo
            </h3>
          </a>
          <div
            class="mt-4"
          >
            <h3
              class="main-footer__subtitle"
            >
              Contact Me
            </h3>
            <p>
              <i
                aria-hidden="true"
                class="fas fa-envelope main-footer__icon"
              />
              <span
                class="main-footer__text"
              >
                vacatamayo321@gmail.com
              </span>
            </p>
          </div>
        </div>
        <div
          class="col-12 col-sm-6 col-xl-3"
        >
          <h3
            class="main-footer__subtitle"
          >
            Social Networks
          </h3>
          <a
            href="https://twitter.com/FernandoVT10"
            target="_blank"
          >
            <p>
              <i
                aria-hidden="true"
                class="fab fa-twitter main-footer__icon"
              />
              <span
                class="main-footer__text"
              >
                @FernandoVT10
              </span>
            </p>
          </a>
          <a
            href="https://github.com/FernandoVT10"
            target="_blank"
          >
            <p>
              <i
                aria-hidden="true"
                class="fab fa-github main-footer__icon"
              />
              <span
                class="main-footer__text"
              >
                FernandoVT10
              </span>
            </p>
          </a>
        </div>
        <div
          class="col-12 col-sm-6 col-xl-3"
        >
          <h3
            class="main-footer__subtitle"
          >
            Description
          </h3>
          <p
            class="main-footer__description"
          >
            This blog is my personal blog xD
          </p>
        </div>
        <div
          class="col-12 col-sm-6 col-xl-3"
        >
          <h3
            class="main-footer__subtitle"
          >
            Suscribe to my blog
          </h3>
          <form>
            <div
              class="formulary__input-container"
            >
              <input
                autocomplete="email"
                class="formulary__input "
                id="-input"
                maxlength="100"
                placeholder="Enter your email"
                required=""
                type="email"
              />
              <label
                class="formulary__error-label mt-2 "
                for="-input"
              >
                <i
                  aria-hidden="true"
                  class="fas fa-info-circle mr-1"
                />
                
              </label>
            </div>
            <button
              class="submit-button mt-3"
            >
              Suscribe
            </button>
          </form>
        </div>
      </div>
    </footer>
    <script
      src="https://code.jquery.com/jquery-3.4.1.slim.min.js"
    />
    <script
      src="https://cdn.jsdelivr.net/npm/popper.js@1.16.0/dist/umd/popper.min.js"
    />
    <script
      src="https://stackpath.bootstrapcdn.com/bootstrap/4.4.1/js/bootstrap.min.js"
    />
    <script
      src="https://kit.fontawesome.com/63ef8f1397.js"
    />
  </div>
</div>
`;
